### PR TITLE
Dsetops feature vfd swmr

### PR DESCRIPTION
**Pull request by vchoi-hdfgroup**

### CODE REVIEW

File: test/vfd_swmr_dsetops_writer.c [hunk 1]
```diff
@@ -0,0 +1,1772 @@
...
+state_init(state_t *s, int argc, char **argv)
+{
+    unsigned long tmp;
+    int           ch;
+    char          tfile[PATH_MAX];
+    char *        end;
+
+    *s = ALL_HID_INITIALIZER;
+    esnprintf(tfile, sizeof(tfile), "%s", argv[0]);
+    esnprintf(s->progname, sizeof(s->progname), "%s", basename(tfile));
```
Comment (kyang2014) [inline test/vfd_swmr_dsetops_writer.c]: May use H5_basename.

File: test/vfd_swmr_dsetops_writer.c [hunk 2]
```diff
@@ -0,0 +1,1772 @@
...
+            case 'm': /* # of rows for -g and/or -k */
+            case 'n': /* # of cols for -g and/or -k */
+            case 's': /* # of sequential writes for -g and/or -k */
+            case 'r': /* # of random writes for -g and/or -k */
+            case 'l': /* # of hyperslab writes for -g and/or -k */
+            case 'w': /* # of raw data modifications for -g and/or -k */
+            case 'u': /* ticks for raeder to wait before verification */
+            case 'c': /* communication interval */
+                errno = 0;
+                tmp   = strtoul(optarg, &end, 0);
```
Comment (kyang2014) [inline test/vfd_swmr_dsetops_writer.c:275]: Use HDF5 string operation like HDstrtoul? Also for the similar string operations?

File: test/vfd_swmr_dsetops_writer.c [hunk 3]
```diff
@@ -0,0 +1,1772 @@
...
+                    TEST_ERROR;
+                }
+            }
+        }
+
+        /* Start verifying raw data modifications for contiguous and/or chunked datasets */
+        if ((s.contig || s.chunked) && s.wwrites) {
+            unsigned k;
+
+            for (step = 0; (step < s.wwrites && step < (s.rows * s.cols)); step++) {
```
Comment (kyang2014) [inline test/vfd_swmr_dsetops_writer.c]: If I understand correctly, by default,  s.rows is 256 and s.cols is 512, for the increasing of each step, the named pipe is used to shake hand, will it take quite a long time to finish the test?

**Review by kyang2014 — COMMENTED** [review]
In vfd_swmr_group_writer.c, I think Ray add the #ifndef H5_HAVE_WIN32_API #else  for future windows tests, maybe it is a good idea for you to add the same #if block.

File: test/vfd_swmr_dsetops_writer.c (hunk 3 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr_dsetops_writer.c]: You are right.  I think I will decrease the default rows and columns to e.g.  10 x 5.

File: test/vfd_swmr_dsetops_writer.c (hunk 2 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr_dsetops_writer.c:275]: Done

File: test/vfd_swmr_dsetops_writer.c (hunk 1 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr_dsetops_writer.c]: Done

File: test/vfd_swmr_dsetops_writer.c [hunk 4]
```diff
@@ -0,0 +1,1771 @@
...
+        }
+    }
+
+    /* Initiailze named pipes */
+    if (s.use_np && !np_init(&np, writer)) {
+        printf("np_init() failed\n");
+        TEST_ERROR;
+    }
+
+    if (writer) {
```
Comment (raylu-hdf) [inline test/vfd_swmr_dsetops_writer.c:1723]: I feel this part could use some subroutines to simplify the code.

File: test/vfd_swmr_dsetops_writer.c [hunk 5]
```diff
@@ -0,0 +1,1771 @@
...
+ *      --dataset with contiguous layout
+ *      --datasets with chunked layout: single/implicit/fixed array/extensible array/btree2 indexes
+ *
+ *      Types of writes:
+ *      -- Sequential writes
+ *      -- Random writes
+ *      -- Regular hyperslab writes
+ *      -- Raw data modifications
+ */
+#include <err.h>
```
Comment (raylu-hdf) [inline test/vfd_swmr_dsetops_writer.c]: These three headers may not be necessary.

Comment (vchoi-hdfgroup) [reply test/vfd_swmr_dsetops_writer.c]: Fixed

File: test/vfd_swmr_dsetops_writer.c (hunk 4 above)
Comment (vchoi-hdfgroup) [reply test/vfd_swmr_dsetops_writer.c:1723]: Fixed.

**Review by kyang2014 — APPROVED** [review]

**Review by raylu-hdf — APPROVED** [review]

**Pull request merged**